### PR TITLE
Fix(Early Stopping): move best score to device

### LINKED ISSUE #7936: Early Stopping throwing error for training on TPUs
## 🐛 Bug

Early Stopping throwing error for training on TPUs

```python
Input tensor is not an XLA tensor
```

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

Pass `EarlyStopping` Callback to Trainer

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #7936 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/early_stopping.py [hunk 1]
```diff
@@ -229,7 +229,7 @@ def _evalute_stopping_criteria(self, current: torch.Tensor) -> Tuple[bool, str]:
                 f" {self.monitor} = {current} {self.order_dict[self.mode]} {self.divergence_threshold}."
                 " Signaling Trainer to stop."
             )
-        elif self.monitor_op(current - self.min_delta, self.best_score):
+        elif self.monitor_op(current - self.min_delta, self.best_score.to(trainer.lightning_module.device)):
```
Comment (justusschock) [inline pytorch_lightning/callbacks/early_stopping.py:232]: TBH: I don't get why this is even necessary. We usually don't remove them from the device (except explicitly specifying this with `move_metrics_to_cpu=True`. So I think we should rather investigate where this happens.

**Review by ethanwharris — APPROVED** [review]
LGTM - at least, it does what it says it does haha

File: pytorch_lightning/callbacks/early_stopping.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/early_stopping.py:232]: Hey @kaushikb11, the tensors are being moved to cpu there: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/logger_connector/result.py#L398.

I think this line was added to enable logger to log values as they need those tensors to be on `cpu`. I guess we could move this function to `log_metrics` in LoggerConnector with apply_to_collection. Mind giving it a try ?

Best,
T.C

Comment (tchaton) [reply pytorch_lightning/callbacks/early_stopping.py:232]: Hey @kaushikb11. Mind giving a try to removing those lines in `ResultCollection.log` ?

Comment (tchaton) [reply pytorch_lightning/callbacks/early_stopping.py:232]: This seems to happen for

which hasn't been moved to the TPU.
 
```
self.best_score = torch_inf if self.monitor_op == torch.lt else -torch_inf
```

Thanks @kaushikb11 for the clarification.

**Review by tchaton — APPROVED** [review]
LGTM !

File: CHANGELOG.md [hunk 2]
```diff
@@ -277,6 +277,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Pass the `stage` argument of `Callback.{setup,teardown}` as a keyword ([#7973](https://github.com/PyTorchLightning/pytorch-lightning/pull/7973))
 
 
+- Fixed move best score to device in EarlyStopping Callback ([#7959](https://github.com/PyTorchLightning/pytorch-lightning/pull/7959))
```
Comment (awaelchli) [inline CHANGELOG.md:280]:
```suggestion
- Fixed move best score to device in EarlyStopping Callback ([#7959](https://github.com/PyTorchLightning/pytorch-lightning/pull/7959))
```
as your comments suggest, this fix applies only for tpu right?

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (kaushikb11) [reply CHANGELOG.md:280]: It applies to other accelerators as well.

Comment (awaelchli) [reply CHANGELOG.md:280]: how would I reproduce an error on a GPU for example?

why can't we do for example this:

`self.best_score.to(current.device)`

moving one tensor to the device but the other not in a monitor_op is only going to raise questions for the reader of this code

Comment (kaushikb11) [reply CHANGELOG.md:280]:
```python
>>> torch_inf = torch.tensor(np.Inf)
>>> value = torch.tensor(5, device=xm.xla_device())
>>> torch.lt(value, torch_inf)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: torch_xla/csrc/aten_xla_bridge.cpp:69 : Check failed: xtensor
*** Begin stack trace ***
	tensorflow::CurrentStackTrace()
```
Similar code for Cuda devices won't throw an error.

Comment (sidml) [reply CHANGELOG.md:280]: I got the following error (device with 2 gpus and using ddp)
`RuntimeError:` Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!`

I also felt that `self.best_score.to(current.device)` was safer..

**Pull request merged**